data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java

```diff
@@ -220,18 +229,19 @@ record = document.getFullDocument().toJson(JSON_WRITER_SETTINGS);
                                 event.delete(DOCUMENTDB_ID_FIELD_NAME);
                                 records.add(event);
                                 recordBytes.add(bytes);
-                                recordCount += 1;
-
-                                if ((recordCount % recordFlushBatchSize == 0) || (System.currentTimeMillis() - lastBufferWriteTime >= bufferWriteIntervalInMs)) {
-                                    LOG.debug("Write to buffer for line {} to {}", (recordCount - recordFlushBatchSize), recordCount);
-                                    writeToBuffer(records, checkPointToken, recordCount);
-                                    lastLocalCheckpoint = checkPointToken;
-                                    lastLocalRecordCount = recordCount;
-                                    lastBufferWriteTime = System.currentTimeMillis();
-                                    bytesProcessedSummary.record(recordBytes.stream().mapToLong(Long::longValue).sum());
-                                    records.clear();
-                                    recordBytes.clear();
+
+                                lock.lock();
```

Comment:
Missing some context. Isn't this worker instance supposed to be owned by a single thread?

Reply:
We have main thread and new thread now. Main loop buffers record and writes when batch size or flush time is reached. The thread writes if main loop has not written for the time duration. The main thread will be blocked if there are no records/updates in stream and it has not flushed the data.